File: data/snippets/snippet.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class Deque:
    def __init__(self):
        self.front = None
        self.back = None

    def is_empty(self):
        return self.front is None

    def push_back(self, data):
        new_node = Node(data)
        if self.back is not None:
            self.back.next = new_node
            new_node.prev = self.back
        else:
            self.front = new_node  
        self.back = new_node

    def push_front(self, data):
        new_node = Node(data)
        if self.front is not None:
            self.front.prev = new_node
            new_node.next = self.front
        else:
            self.back = new_node  
        self.front = new_node

    def pop_back(self):
        if self.is_empty():
            return None
        data = self.back.data
        self.back = self.back.prev
        if self.back is None:
            self.front = None  
        else:
            self.back.next = None
        return data

    def pop_front(self):
        if self.is_empty():
            return None
        data = self.front.data
        self.front = self.front.next
        if self.front is None:
            self.back = None  
        else:
            self.front.prev = None
        return data

    def back_value(self):
        if self.is_empty():
            return None
        return self.back.data

    def front_value(self):
        if self.is_empty():
            return None
        return self.front.data

    def clear(self):
        self.front = None
        self.back = None

    def size(self):
        count = 0
        current = self.front
        while current:
            count += 1
            current = current.next
        return count

if __name__ == "__main__":
    deque = Deque()
    deque.push_back(1)
    deque.push_back(2)
    deque.push_front(0)

    print("Front:", deque.front_value())  
    print("Back:", deque.back_value())    

    print("Popped from front:", deque.pop_front())  
    print("Popped from back:", deque.pop_back())    

    print("Size after pops:", deque.size())  

File: data/snippets/snippet.cpp
#include <iostream>
#include <memory>

template<typename T>
class Deque {
private:
    struct Node {
        T data;
        std::shared_ptr<Node> next;
        std::shared_ptr<Node> prev;

        Node(T value) : data(value), next(nullptr), prev(nullptr) {}
    };

    std::shared_ptr<Node> front;
    std::shared_ptr<Node> back;

public:
    Deque() : front(nullptr), back(nullptr) {}

    ~Deque() {
        clear();
    }

    void pushBack(const T& item) {
        auto newNode = std::make_shared<Node>(item);
        if (back) {
            back->next = newNode;
            newNode->prev = back;
        } else {
            front = newNode; // If deque is empty, new node is also the front
        }
        back = newNode;
    }

    void pushFront(const T& item) {
        auto newNode = std::make_shared<Node>(item);
        if (front) {
            front->prev = newNode;
            newNode->next = front;
        } else {
            back = newNode; // If deque is empty, new node is also the back
        }
        front = newNode;
    }

    std::shared_ptr<T> popBack() {
        if (isEmpty()) {
            return nullptr;
        }

        auto dataPtr = std::make_shared<T>(back->data);
        back = back->prev;

        if (back) {
            back->next = nullptr;
        } else {
            front = nullptr; // If deque is now empty
        }

        return dataPtr;
    }

    std::shared_ptr<T> popFront() {
        if (isEmpty()) {
            return nullptr;
        }

        auto dataPtr = std::make_shared<T>(front->data);
        front = front->next;

        if (front) {
            front->prev = nullptr;
        } else {
            back = nullptr; // If deque is now empty
        }

        return dataPtr;
    }

    std::shared_ptr<T> backValue() const {
        if (isEmpty()) {
            return nullptr;
        }
        return std::make_shared<T>(back->data);
    }

    std::shared_ptr<T> frontValue() const {
        if (isEmpty()) {
            return nullptr;
        }
        return std::make_shared<T>(front->data);
    }

    void clear() {
        front = nullptr;
        back = nullptr;
    }

    bool isEmpty() const {
        return front == nullptr;
    }

    int size() const {
        int count = 0;
        auto current = front;
        while (current) {
            count++;
            current = current->next;
        }
        return count;
    }
};

int main() {
    Deque<int> deque;
    deque.pushBack(1);
    deque.pushBack(2);
    deque.pushFront(0);

    std::cout << "Front: " << *deque.frontValue() << std::endl; // Output: 0
    std::cout << "Back: " << *deque.backValue() << std::endl;   // Output: 2

    std::cout << "Popped from front: " << *deque.popFront() << std::endl; // Output: 0
    std::cout << "Popped from back: " << *deque.popBack() << std::endl;   // Output: 2

    std::cout << "Size after pops: " << deque.size() << std::endl; // Output: 1

    return 0;
}